spec/app/ics-100-atomic-swap/README.md

```diff
@@ -0,0 +1,593 @@
+
+### Data Structures
+
+Only one packet data type is required: `AtomicSwapPacketData`, which specifies the swap message type, data (protobuf marshalled) and a memo field.
+
+```typescript
+enum SwapMessageType {
+  // Default zero value enumeration
+  TYPE_UNSPECIFIED = 0,
+  TYPE_MSG_MAKE_SWAP = 1,
+  TYPE_MSG_TAKE_SWAP = 2,
+  TYPE_MSG_CANCEL_SWAP = 3,
+}
+```
+
+```typescript
+// AtomicSwapPacketData is comprised of a swap message type, raw transaction and optional memo field.
+interface AtomicSwapPacketData {
+  type: SwapMessageType;
+  data: []byte;
+  memo: string;
+}
+```
+
+All `AtomicSwapPacketData` will be forwarded to the corresponding message handler to execute according to its type. There are 3 types:
+
+```typescript
+interface MakeSwapMsg {
+  // the port on which the packet will be sent, specified by the maker when the order is created
+  sourcePort string;
+  // the channel on which the packet will be sent, specified by the maker when the order is created
+  sourceChannel: string;
+  // the tokens to be exchanged
+  sellToken : Coin;
+  buyToken: Coin;
+  // the maker's address
+  makerAddress: string;
+  // the maker's address on the taker chain
+  makerReceivingAddress string;
+  // if desiredTaker is specified,
+  // only the desiredTaker is allowed to take this order
+  // this is the address on the taker chain
+  desiredTaker: string;
+  creationTimestamp: int64;
+  expirationTimestamp: int64;
+  timeoutHeight: int64,
+  timeoutTimestamp: int64,
+}
+```
+
+```typescript
+interface TakeSwapMsg {
+  orderId: string;
+  // the tokens to be sell
+  sellToken: Coin;
+  // the taker's address
+  takerAddress: string;
+  // the taker's address on the maker chain
+  takerReceivingAddress: string;
+  creationTimestamp: int64;
+  timeoutHeight: int64,
+  timeoutTimestamp: int64,
+}
+```
+
+```typescript
+interface TakeCancelMsg {
+  orderId: string;
+  makerAddress: string;
+  timeoutHeight: int64,
+  timeoutTimestamp: int64,
+}
+```
+
+Note: Coin is sdk.Coin, 
+```ts
+interface Coin {
+  amount: int64,
+  denom: string
+}
+```
+
+Both the maker chain and taker chain maintain separate orderbooks. Orders are saved in both maker chain and taker chain.
+
+```typescript
+enum Status {
+  INITIAL = 0,
+  SYNC = 1,
+  CANCEL = 2,
+  COMPLETE = 3,
+}
+
+interface OrderBook {
+  id: string;
+  maker: MakeSwapMsg;
+  status: Status;
+  // set onRecieved(), Make sure that the take order can only be sent to the chain the make order came from
+  portIdOnTakerChain: string;
+  // set onRecieved(), Make sure that the take order can only be sent to the chain the make order came from
+  channelIdOnTakerChain: string;
+  taker: TakeSwap;
+  cancelTimestamp: int64;
+  completeTimestamp: int64;
+  
+  createOrder(msg: MakeSwapMsg) OrderBook {
+    return {
+        id : generateOrderId(msg)
+        status: Status.INITIAL
+        maker: msg,
+    }
+  }
+}
+
+// Order id is a global unique string
+function generateOrderId(msg MakeSwapMsg) {
```

Comment:
This seems like it will collide if I make two separate but identical maker orders in the same block

Reply:
add a `nonce`  in make message.

Reply:
Why can't you just use the packet sequence? And remove the nonce

Reply:
>  Why can't you just use the packet sequence? And remove the nonce
No specific reasons, both are work.  

I am going to move `orderPath` into the Order.  and `generateOrderId` take the orderPath as argument. seems we can have universal store key.